addons/test_mail_full/tests/test_portal.py

```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+                break
+        else:
+            raise AssertionError('Record access URL not found')
+        # Transform that access URL to have one with an invalid access token
+        cls.record_access_url_wrong_token = re.sub(r'access_token=[0-9a-z-]*',
+                                                   'access_token=00000000-0000-0000-0000-000000000000',
+                                                   cls.record_access_url)
+
+    def test_initial_data(self):
+        """Test some initial value."""
+        self.assertNotEqual(self.record_access_url, self.record_access_url_wrong_token)
+
+    def _get_composer_with_context(self, template_id=False):
+        return self.env['mail.compose.message'].with_context({
+            'default_model': self.record_customer._name,
+            'default_res_id': self.record_customer.id,
+            'default_use_template': template_id,
+            'default_template_id': template_id,
+            'default_composition_mode': 'comment',
+            'mark_so_as_sent': True,
+            'default_email_layout_xmlid': 'mail.mail_notification_layout_with_responsible_signature',
+            'force_email': True,
+            'model_description': 'Quotation',
+        })
+
+    @users('employee')
+    def test_send_message_to_customer_using_template(self):
+        """Send a mail to a customer without an account and check that it contains a link to view the record.
+
+        Other tests below check that that same link has the correct behavior.
+        This test follows the common use case by using a template while the next send the mail without a template."""
+        composer = self._get_composer_with_context(self.mail_template.id).create({})
+        # currently onchange necessary
+        composer._onchange_template_id_wrapper()
+
+        with self.mock_mail_gateway(mail_unlink_sent=True):
+            composer._action_send_mail()
+
+        self.assertEqual(len(self._mails), 1)
+        self.assertIn(html_escape(self.record_access_url), self._mails[0].get('body'))
+        self.assertEqual('Your quotation', self._mails[0].get('subject'))  # Check that the template is used
+
+    @users('employee')
+    def test_send_message_to_customer(self):
+        """Same as test_send_message_to_customer_using_template but without a template."""
+        composer = self._get_composer_with_context().create({
+            'subject': 'Quotation',
+            'body': '<p>Hello, your quotation is ready for review.</p>',
+            'partner_ids': self.customer.ids,
+        })
+
+        with self.mock_mail_gateway(mail_unlink_sent=True):
+            composer._action_send_mail()
+
+        self.assertEqual(len(self._mails), 1)
+        self.assertIn(html_escape(self.record_access_url), self._mails[0].get('body'))
+        self.assertEqual('Quotation', self._mails[0].get('subject'))  # Check that the template is not used
+
+    def test_customer_access_not_logged(self):
```

Comment:
Side note: could you order tests alphabetically in a given class ? Same reason as for classes: easier when doing refactoring, having tests following the file order, better when having several tests to fix or check :D

Reply:
Done, the only test that I didn’t move is test_initial_data because I think it should be the first test. What do you think?

Reply:
That's ok :) we should find a synonym beginning with a :thinking: